Fix ecliptic transformations

Juanlu001 opened the pull request:
SOFA/ERFA new functions for ecliptic transformations do not use the nutation matrix (see https://github.com/astropy/astropy/issues/4873#issuecomment-326528463). Adjusting the corresponding Astropy function seemed to fix the precision issues with respect to the rest of the tools.

If this is merged, I think there is no need to go for #4873.

Before/After:

![before](https://user-images.githubusercontent.com/316517/30005055-7228aeb2-90da-11e7-9ea7-204bdf49b4b0.png)

![after](https://user-images.githubusercontent.com/316517/30005057-7884fb30-90da-11e7-981e-3112599b3a6c.png)

astrofrog commented:
Thanks! If this change is correct (I'll leave that decision up to @eteq and others), this needs a comment when calling the ERFA function to explain the choice of ERFA function (so people don't change it back), and a changelog entry.

Juanlu001 commented:
Some of the failures are due to timeouts or jobs not starting, but some others are doctest precision differences with `get_body_barycentric('moon', t, ephemeris='de432s')`. I am not sure how those are related to this PR though.

Talking about the choice of ERFA function, in the absence of nutation I am not sure if we can keep calling these systems `TrueEcliptic`. Perhaps `MeanEcliptic` would be more appropriate @eteq?

astrofrog commented:
I wonder if this means we should actually *add* frames called ``MeanEcliptic`` and leave the current ones as-is, and then use the mean ones for the coordinates benchmark?

Juanlu001 commented:
I added yet another correction that I have been working on for several weeks now and that was causing us great pain in poliastro.

On the one hand, the vector arithmetic for the computation of the position of the Sun was flawed. I attach a sketch to support the corresponding change.

![file1504542008454](https://user-images.githubusercontent.com/316517/30034839-6e98eb5c-91a4-11e7-97cf-34585db38a87.jpg)

On the other hand, the `equinox` attribute of `HeliocentricTrueEcliptic` was being used both to compute the obliquity and the position of the Sun. I left the former, but for the latter I used a new attribute, `obstime`, consistent with what it's done with GCRS.

Thanks to these changes, we managed to read data from the JPL Small-Body Database Browser, set it to the corrected `HeliocentricTrueEcliptic`, convert it to ICRS using Astropy machinery and retrieve the same results that HORIZONS gives. You can imagine how proud we are! You can see our work in progress results here:

http://nbviewer.jupyter.org/github/Juanlu001/poliastro/blob/florence-example/docs/source/examples/Catch%20that%20asteroid!.ipynb

Surprisingly (at least on first thought), the coordinates-benchmarks results did not change. I am not sure about the reasons for this, but perhaps it is only checking the coordinates of very distant objects and traslations of ~800 000 km are not significant (whereas small changes in rotations have dramatic impacts). I would like to ask for guidance regarding this specific issue.

If the devs give their OK from an algorithmic point of view, I will continue adding tests, documentation and a changelog entry.

mhvk commented:
@Juanlu001 - this looks very promising indeed!  I have no time to look at this myself (and am also not a real expert anyway), but also @StuartLittlefair, who wrote quite a bit of these parts of the code.

StuartLittlefair commented:
@Juanlu001 these fixes look great, and look right to me.

My only questions are about frame naming, and the separate ```equinox``` and ```obstime``` parameters.

With frame naming, ecliptic frames normally use either a mean equinox of a standard date (i.e J2000), the mean equinox of a specific date (i.e what you have here) or the true equinox of a specific date (precession + nutation). We definitely shouldn't call this ```TrueEcliptic``` any more! My preference is to add your barycentre fix to the ```TrueEcliptic``` frame and add a new ```MeanEcliptic``` frame which doesn't have nutation in it.

With respect to using separate ```equinox``` and ```obstime``` parameters I'm not sure I agree with what you've done here. The ```equinox``` of the frame should specify the position *and* obliquity of the Sun IMO. Given that, I'm not sure of the utility of a seperate ```obstime``` parameter. Perhaps @eteq can clarify, since he understands the distinction better than I.

Juanlu001 commented:
> My preference is to add your barycentre fix to the `TrueEcliptic` frame and add a new `MeanEcliptic` frame which doesn't have nutation in it.

Cool, will do that!

> With respect to using separate `equinox` and `obstime` parameters I'm not sure I agree with what you've done here. The `equinox` of the frame should specify the position _and_ obliquity of the Sun IMO.

I don't have an opinion at all on how these frames _should_ be, so I tried to replicate what NASA and JPL do in their public APIs to suit my needs. In my view (after much struggling), `Heliocentric*` frames should be equivalent to Geocentric frames in this respect - having an `osbtime` attribute to tell the position of the central body, be it the Earth or the Sun. On the other hand, I cannot remove the `equinox` since the value for the obliquity is fixed in J2000 and not in the true date as I copied here:

https://github.com/astropy/astropy/issues/6461#issuecomment-326573318

I don't have a problem in removing `obstime` and using `equinox` for finding the position of the Sun, but that will render these systems unusable for me and I will have to redefine them.

StuartLittlefair commented:
@Juanlu001 

Sorry for the slow conversation; things are extremely hectic for me at the moment. If I understand correctly then, you wish for the '''MeanEcliptic''' frame to use the same epoch for the obliquity as Horizons, but a user-specified epoch for the central body's position? However, there's no such requirement for the '''TrueEcliptic''' frame?

Would the frames be useable if they had one parameter (e.g '''obstime'''), and the TrueEcliptic frame used this for both the obliquity and the position of the Sun, but the MeanEcliptic frame used a reference epoch (not given as a parameter, but documented) for the position and/or obliquity? If not, perhaps if you can show a use case that would fail it would help me understand why both '''equinox''' and '''obstime''' are needed.

Juanlu001 commented:
> Sorry for the slow conversation; things are extremely hectic for me at the moment.

No need to apologize :)

> If I understand correctly then, you wish for the '''MeanEcliptic''' frame to use the same epoch for the obliquity as Horizons, but a user-specified epoch for the central body's position?

Correct. In fact, I don't even _need_ a variable epoch for the obliquity - they just use the value of J2000 always. What I do need though is a way to specify the position of the Sun.

> However, there's no such requirement for the '''TrueEcliptic''' frame?

I have no way to validate this. My entire journey has been to replicate what [JPL Horizons](https://ssd.jpl.nasa.gov/horizons.cgi) gives (you can do the experiments by yourself), and the web interface only has one ecliptic frame to select:

>   Ecliptic and Mean Equinox of Reference Epoch
> 
>     Reference epoch: J2000.0
>     XY-plane: plane of the Earth's orbit at the reference epoch
>               Note: obliquity of 84381.448 arcseconds wrt ICRF equator (IAU76)
>     X-axis  : out along ascending node of instantaneous plane of the Earth's
>               orbit and the Earth's mean equator at the reference epoch
>     Z-axis  : perpendicular to the xy-plane in the directional (+ or -) sense
>               of Earth's north pole at the reference epoch.

Probably because it's the only modern ecliptic frame defined in SPICE:

https://naif.jpl.nasa.gov/pub/naif/toolkit_docs/C/req/frames.html#Appendix.%20%60%60Built%20in%27%27%20Inertial%20Reference%20Frames

There were already concerns about validation of ecliptic frames (see original issue https://github.com/astropy/astropy/issues/6461 and discussion in https://github.com/astropy/astropy/pull/3749). This pull request can be considered another data point.

> Would the frames be useable if they had one parameter (e.g '''obstime'''), and the TrueEcliptic frame used this for both the obliquity and the position of the Sun, 

I won't use the `TrueEcliptic` so I'm not very concerned about this :)

> but the MeanEcliptic frame used a reference epoch (not given as a parameter, but documented) for the position and/or obliquity?

Yes, definitely. If we have a reference epoch (say, J2000) for the obliquity and an `obstime` parameter for the position of the Sun, this would work perfectly for me. If you agree I can implement exactly this, and leave `TrueEcliptic` with just one parameter.

StuartLittlefair commented:
Ok, excellent. I would keep the single parameter's name the same as before to minimise API changes. So I would implement two frames, and have '''equinox''' set the position and obliquity in '''TrueEcliptic''' but only the position in '''MeanEcliptic'''

mhvk commented:
The name `MeanEcliptic` seems the right one, but I am less sure about using `equinox` to then define the position of the Sun, especially as in `HCRS` the position of the Sun is determined from `obstime` (as noted by @Juanlu001 above).

But perhaps I should first ensure I really understand what these coordinates are:
* For `BarycentricMeanEcliptic`, only the equinox matters, as this gives the obliquity and thus the orientation of the coordinates relative to ICRS, defining the coordinate system fully, as there are offsets involved.
* For `Helio/GeocentricMeanEcliptic`, the obliquity defines the plane, but the offset within that plane depends still on the time via the positions of the Sun and Earth. 
* For `BarycentricTrueEcliptic`, the obliquity now depends on time, so this has variable orientation (but no offset).
* For `Helio/GeocentricTrueEcliptic`, both orientation and offset now depend on time.

With this list, it seems fairly obvious that `equinox` has to be the time when the obliquity is taken at, as it defines orientation. However, offsets in other frames like `GCRS` and `HCRS` are inferred from `obstime`, so to me that seems most logical here too, i.e., I would thus advocate the `*MeanEcliptic` frames to have both `equinox` and `obstime` (i.e., they become what @Juanlu001 implemented here).

For the `*TrueEcliptic` frames, it would seem that one has to insist that `obstime` and `equinox` are in fact identical (which I think would need to be implemented).

Separately, looking at the links @Juanlu001 posted, the splice reference frames include:
```
      17  ECLIPJ2000  Ecliptic coordinates based upon the
                      J2000 frame.
 
                      The value for the obliquity of the
                      ecliptic at J2000 is taken from page 114
                      of [7] equation 3.222-1. This agrees with the
                      expression given in [5].
 
      18  ECLIPB1950  Ecliptic coordinates based upon the B1950
                      frame.
 
                      The value for the obliquity of the ecliptic at
                      B1950 is taken from page 171 of [7].
```
It seems the frame actually wanted here is the equivalent of `ECLIPJ2000`, i.e., it has a fixed equinox. I'm not quite sure how `ECLIPB1950` would fit in here, as it changes both obliquite and reference frame for that obliquity.

paulray commented:
@mhvk, you might also want input from @demorest and David Nice (who may not be on github)

Juanlu001 commented:
Just to clarify the terminology @mhvk: acoording to the appendix of the [SOFA Tools for Earth Attitude cookbook](http://www.iausofa.org/sofa_pn_c.pdf), the terms "true" and "mean" refer to the inclusion of the nutation terms in the rotation matrix. As @eteq explained in his original pull request (see https://github.com/astropy/astropy/pull/3749#issuecomment-143902959), he ended up adding the "true" versions and not the "mean" versions.

Fast forward until two weeks ago, I realized that neither `HeliocentricTrueEcliptic` nor `BarycentricTrueEcliptic` were suiting my needs. In the beginning I thought it was a problem of accounting for the nutation (see https://github.com/astropy/astropy/issues/4873#issuecomment-326528463), so I fixed it. The effects of this were:

* The coordinates-benchmark numbers improved by several orders of magnitude. This means that the other tools we are comparing to are not accounting for the nutation either. Notice that there is no way to know which precession model they are using (if any) since, at J2000, pre and post IAU 2000 conventions agree (by design).
* These frames can no longer be called `*TrueEcliptic` since, as per the standard naming, they don't account for nutation anymore.
* My problems were still there. When computing the position of the Florence asteroid and comparing it to the Horizons system, I was having errors "of the same order of magnitude than the norm of the position vector of the Sun with respect to the Solar System Barycenter" (see https://github.com/poliastro/poliastro/issues/222#issuecomment-326822170) and the nutation change was of little effect.

So I went on investigating, and found that the position of the Sun was the main driver of the errors - in Astropy it was being computed using the `equinox` parameter (default to J2000), but apparently in NASA/SPICE/Horizons it was computed at the observation time, independently from the value of the obliquity. I fixed the computation of the position of the Sun, and the effects of this were:

* Absolutely no changes in coordinates-benchmark - not a single decimal digit. I still wonder why, but I guess it has to do with the fact that origin offsets do not affect much angular observations.
* The result of `fetch Florence position from NASA in "their" ecliptic frame + assign it to our NewHeliocentricMeanEcliptic + convert to Astropy ICRS` was exactly equal to `fetch Florence position from NASA in "their" ICRS`, so I called it a day.

After all this journey, apart from being precise with the `True/Mean` naming, I don't have a strong opinion on whether `equinox` and/or `obstime` should be used to set the position of the Sun (center of the Heliocentric frames) and the plane of the ecliptic (be it `mean of date`, `mean of epoch`, `true of date` or `true of epoch`). I _feel_ that using equinox for everything might be a bit confusing, but of course it's a minor issue and I would not complain about that.

Of course, by fixing _my_ needs I don't want to screw up the rest of the folks that use this, so I'm happy they are all brought to the conversation :)

mhvk commented:
@Juanlu001 - Thanks for the summary! Reading back the comments in #3749, it seems to me that we indeed want the `Mean` variants and that those should have both `equinox` and `obstime` for `Geo/Helio`. For the `True` variants, I think it simply means `equinox` is always equal to `obstime`.

But really @eteq should pipe in as well..

mhvk commented:
Separately, shouldn't `Heliocentric*Ecliptic` connect to `HCRS`, just like `Geocentric*Ecliptic` connects to `GCRS`? (and then any future B1950 ecliptic coordinate system would connect to `FK4`...).

Juanlu001 commented:
> Separately, shouldn't Heliocentric*Ecliptic connect to HCRS, just like Geocentric*Ecliptic connects to GCRS?

+1!

StuartLittlefair commented:
@mhvk @Juanlu001 I think I'm clearer on the issue here now. I think it's fair to say that the only difference between ```TrueEcliptic``` and ```MeanEcliptic``` frames is whether one includes nutation or not.

On reflection, I can see no problem with including ```equinox``` and ```obstime``` parameters for mean frames, but perhaps they need very carefully documenting of what each parameter does, since these terms are different to the "epoch" and "date" terminology usually adopted for ecliptic frames?

It is probably also worth documenting somewhere which equinox/epoch to use to get the same results as Horizons....

Juanlu001 commented:
> I think it's fair to say that the only difference between TrueEcliptic and MeanEcliptic frames is whether one includes nutation or not.

Correct!

> On reflection, I can see no problem with including equinox and obstime parameters for mean frames, but perhaps they need very carefully documenting of what each parameter does, since these terms are different to the "epoch" and "date" terminology usually adopted for ecliptic frames?

In fact, "of epoch" equals "of date when date is J2000", at least in the books and papers I have read (perhaps others think otherwise).

Anyway, I agree with documenting this very carefully, it's a confusing topic and some pointers to "ground truth" references (like the SOFA cookbook, also [USNO Circular 179](http://aa.usno.navy.mil/publications/docs/Circular_179.pdf)) are totally worth it.

> It is probably also worth documenting somewhere which equinox/epoch to use to get the same results as Horizons....

With these changes, we can note in the docstring of `HeliocentricMeanEcliptic` that it is the closest one to what Horizons uses. However, for utter precision we should also remove the precession, since they use a pure rotation around the X axis. Using the [SpiceyPy](http://spiceypy.readthedocs.io/en/master/index.html) library to directly read the NASA SPICE kernels and [compute the rotation matrix](https://naif.jpl.nasa.gov/pub/naif/toolkit_docs/C/cspice/pxform_c.html):

```
In [17]: rot = spice.pxform('ECLIPJ2000', 'J2000', et)  # After loading the kernels, etc

In [18]: rot
Out[18]: 
array([[ 1.        ,  0.        ,  0.        ],
       [ 0.        ,  0.91748206, -0.39777716],
       [ 0.        ,  0.39777716,  0.91748206]])

In [19]: np.arccos(rot[1, 1]) * u.rad
Out[19]: <Quantity 0.40909280422232897 rad>

In [20]: _.to(u.arcsec)
Out[20]: <Quantity 84381.448 arcsec>
```

Which is the value that `_erfa.obl80` gives:

```
In [14]: equinox = Time("J2000", scale="tdb")

In [15]: jd1, jd2 = get_jd12(equinox, 'tdb')

In [16]: obl80 = _erfa.obl80(jd1, jd2) * u.radian

In [17]: obl80.to(u.arcsec)
Out[17]: <Quantity 84381.448 arcsec>

In [18]: obl = _erfa.obl06(jd1, jd2) * u.radian

In [19]: obl.to(u.arcsec)
Out[19]: <Quantity 84381.406 arcsec>
```

I understand however that this might be too application-specific. If we do not include it in Astropy I will include it in poliastro anyway :)

mhvk commented:
@Juanlu001 - in principle, it seems to me that if it is used in the SPICE kernels, astropy should support it, or at least we should document how to get it. Just to be sure, in the frame defined for SPICE, which is J2000, both precession and nutation should be irrelevant, correct? Hence, the only term left is the "frame bias" - so is it that term that is not applied in `ECLIPJ2000`? It might still be good to define corresponding ecliptic frames, though I'm a bit unsure about the names. Maybe `*centricJ2000Ecliptic`? This then would have no `equinox`, but still have an `obstime`.

Just as another reference point, looking at PINT's [`pulsar_ecliptic.py`]( https://github.com/nanograv/PINT/blob/4dbcab90b603c06f1e3e5605bfaab83d458efbcf/pint/pulsar_ecliptic.py#L53), I see that their (barycentric) ecliptic frame also implements a transformation to ICRS that worries about obliquity only, i.e., that ignores the frame bias. (Not sure if that is in fact intended; @paulray, @demorest?)

demorest commented:
@mhvk -- I haven't had time to really read/understand this discussion fully yet.  As far as PINT's implementation, the only intent was to be compatible with how TEMPO/TEMPO2 have traditionally defined ecliptic coordinates.  As you say, this is a simple rotation using obliquity only; there are several choices for what obliquity value is used, there may be some useful historical info in $TEMPO/ephem/ecliptic.dat.

Juanlu001 commented:
> Just to be sure, in the frame defined for SPICE, which is J2000, both precession and nutation should be irrelevant, correct?

To be clear: for the transformation between J2000 and ECLIPJ2000 either way, both precession and nutation are irrelevant.

> Hence, the only term left is the "frame bias" - so is it that term that is not applied in ECLIPJ2000?

I think that correction was introduced in the new IAU resolutions, and from what I understand they are using the old models.

Juanlu001 commented:
> It might still be good to define corresponding ecliptic frames, though I'm a bit unsure about the names. Maybe *centricJ2000Ecliptic? This then would have no equinox, but still have an obstime.

My proposal:

* `BarycentricTrueEcliptic` & `HeliocentricTrueEcliptic` with IAU 2006 precession, nutation and obliquity of `equinox` (default to J2000 TDB) + `obstime` parameter for the position of the Sun for `Heliocentric` only. These ones should **not** be used in coordinates-benchmark, and it should be noted that there are no sources for validation<sup>1</sup>.
* `BarycentricMeanEcliptic` & `HeliocentricMeanEcliptic` with IAU 2006 precession and obliquity of `equinox` (default to J2000 TDB) + `obstime` parameter for the position of the Sun (`Heliocentric` only). These ones **should** be used in coordinates-benchmark<sup>2</sup>.
* `EclipticJ2000` with IAU 1980 precession<sup>3</sup> and obliquity of J2000 (no equinox parameter), centered in the Sun with `obstime` parameter, explicitly for Horizons compatibility. This can be validated copying values from SpiceyPy and hardcoding them in the tests.

<sup>1</sup>In principle I would be against including frames that cannot be externally validated, but I understand that maybe others want them there anyway.
<sup>2</sup>It is possible that the other libraries in coordinates-benchmark use the IAU 1976 precession (see <sup>3</sup>) and IAU 1980 obliquity, in which case `EclipticJ2000` would perform better. It should be straightforward to check that.
<sup>3</sup>The IAU 1976 precession matrix at J2000 is the identity:

```
In [13]: equinox = Time("J2000", scale="tdb")

In [14]: jd1, jd2 = get_jd12(equinox, 'tdb')

In [15]: _erfa.pmat76(jd1, jd2)
Out[15]: 
array([[ 1.,  0.,  0.],
       [ 0.,  1.,  0.],
       [ 0.,  0.,  1.]])
```

Does this make sense?

Juanlu001 commented:
(For those following this discussion by email -- I made some tweaks to my last comment, please open GH issue)

mhvk commented:
> I think that correction was introduced in the new IAU resolutions, and from what I understand they are using the old models.

OK, that explains things. The same would hold for `tempo/tempo2`, which well predate the IAU 20* resolutions.  And the fact that both PINT and SPICE are using this argues that we should include them.

...

Just seeing the above comment: yes, I agree with the three categories (am not sure we actually want/need any of the `TrueEcliptic`, but should keep them for backward compatibility; it would need `obstime=equinox` by default).

For the third, we'd need both barycentric and heliocentric versions (for PINT, SPICE, resp.), and I think the obliquity might as well be a `QuantityAttribute` so `PINT` can vary it. It does seem we should either have *no* precession or we should allow an equinox (for the 1976 precession). If we go with the latter, my tendency would be to call those `EclipticIAU1976` or something similar. An advantage of that might be that it more clearly indicates that this is not the preferred solution (as might be guessed from a `J2000` name). Alternatively, if we don't want to give the impression that the precession matters, it could be something like `EclipticNoFrameBias` (in analogy with `FK4NoETerms`). Not sure what would work best; ideally, we keep open the option to have an `FK4/B1950` variant... Most important will be to document this...

p.s. Since we're now focused on names, it seems like we are converging!

Juanlu001 commented:
> I think the obliquity might as well be a QuantityAttribute so PINT can vary it.

I guess you refer to this:

https://github.com/nanograv/PINT/blob/4dbcab9/pint/pulsar_ecliptic.py
https://github.com/nanograv/PINT/blob/4dbcab90b603c06f1e3e5605bfaab83d458efbcf/pint/datafiles/ecliptic.dat

Specifying the obliquity from the outside conflicts with computing it internally using some `equinox` and its `jd1` and `jd2`. I am not sure how such an implementation could be.

> It does seem we should either have no precession or we should allow an equinox (for the 1976 precession)

Of the people that have chimed in so far, I think nobody asked for an Ecliptic frame with an arbitrary equinox. The PINT folks do want different obliquity values, but they all refer to J2000 - they differ because of the underlying model, not because of the date. Two of these models are available in `_erfa` (namely IAU 2006 and 1976/1980) but some other models are not (DE 403, IERS 2003).

Perhaps instead of providing a custom equinox, it is more useful for most people fixing the equinox to J2000 (not even a default parameter, just fix it) and be able to configure the precession/nutation/obliquity model from the outside.

With this idea in mind, three cases arise:

1. **IAU 2006**: the value is 84381.406000 arcsec, which can be obtained using `_erfa.obl06` at J2000. Precession + frame bias of J2000 must be computed (although PINT is not taking these corrections into account, only rotating `OBL` around X), and also nutation for the `True` versions (although they cannot be validated).
2. **IAU 1976/1980**: the value is 84381.448 arcsec, which can be obtained using `_erfa.obl80` at J2000. No precession + frame bias correction is required (which is consistent with what PINT does), and applying nutation for the `True` version without precession seems weird, and probably wrong.
3. Other models that predate IAU 2006 and cannot be computed in ERFA. Should not apply precession + frame bias either (again consistent with PINT). For the same reasons, nutation does not make sense.

I already feel I'm being extremely verbose (boring?), but as a last note I would like to say that providing too many reference frames might confuse potential uses (this topic is already difficult, so imagine offering six or eight potential options with similar names...). I would save the discard the ones that we cannot validate or proof useful, and have two, three or at most four:

* Mean IAU 2006 of J2000 (modern systems)
* Mean IAU 1976 of J2000 (Horizons, others)
* (Optional) Mean arbitrary obliquity (weird systems like tempo/tempo2, IAU 2006 without precession + frame bias)
* (Very Optional) True IAU 2006 of equinox (unused?)

Notice that **all these systems** need to correctly compute the position of the Sun in a consistent way.

If backwards compatibility is important, then we could turn "Mean IAU 2006 of J2000" in "Mean IAU 2006 of equinox" and keep the (unused?) "True IAU 2006 of equinox". But they were marked as experimental anyway, so perhaps it's not a big deal.

Juanlu001 commented:
> Specifying the obliquity from the outside conflicts with computing it internally using some equinox and its jd1 and jd2. I am not sure how such an implementation could be.

I meant: I am not sure how we can conceal both ways of specifying the rotation at the same time.

Juanlu001 commented:
> I would discard the ones that we cannot validate or proof useful, and have two, three or at most four:

Multiplied by two: Heliocentric and Barycentric...

mhvk commented:
@Juanlu001 - I think I'm with you on your final recommendation, just `MeanEcliptic` systems for IAU 1976/80 and IAU 2006. I'm also fine with not allowing an equinox (i.e., it is simply fixed); we can always add it if there is actual use for it. Given that they seem to be the only systems in actual use, it may be an idea to just remove `Mean` from the name. I also agree we should remove or deprecate the `TrueEcliptic` systems.

But note we still end up with *six* classes: `Geocentric`, `Heliocentric`, and `Barycentric` variants. It seems that `Barycentric` will now have *no* attributes, while `HelioCentric` and `Geocentric` will need `obstime`.

StuartLittlefair commented:
Do we really need six classes, or can the model for the obliquity be a frame attribute?

astrofrog commented:
I'm with @StuartLittlefair here - six classes seems a bit over the top, and a frame attribute might make more sense for the obliquity.

Juanlu001 commented:
Either one is fine for me. For choosing the obliquity, do you think it would be better to have a text parameter like `obliquity="IAU06"` vs `obliquity="IAU80"` or rather accept the numerical value?

mhvk commented:
Precedent for `FK4` (with `NoETerms`) might suggest different classes. My own sense would be that one should have a different class if the actual transformation steps are different, while one can use an attribute if the transformation depends on specific values.

Perhaps more to the point, it should be possible for the frame transformation graph to figure out how to transform from one IAU standard to the other. I think this is most easily done via classes (it would presumably go through ICRS rather than directly from one to the other).

Juanlu001 commented:
In case anyone is interested, here's what I did in poliastro:

Juanlu001 commented:
In case anyone is interested, this is what I did in poliastro:

https://github.com/poliastro/poliastro/blob/0.7.x/src/poliastro/frames.py

I won't be able to finish this pull request until the end of the month, please ping me in October if I haven't said anything here :)

eteq commented:
Just so that no one things I'm ignoring this... it is clearly worthy of *deep* thought/attention, and I had been trying not to give an opinion until I've had time to think deeply. But I've gotten *part* way there now, and here is  my current (although not finalized) thinking:

* I'm tilting towards @mhvk's point that these all need to be distinct frames because they really *are* different transformations.  But I also have the misgivings about adding so many more frames so am trying to work out whether there's any way to use attributes to set some of this without violating @mhvk's guidelines.
* I am *not* comfortable with dropping the `equinox` attributes completely. Suppose I actually care about the *actual* ecliptic at a specific instant in time because I care about where things are relative to the actual Earth's orbit at that instant.  That requires obstime and equinox to be the same.  So I believe that means adding @Juanlu001's "Very Optional" case.  The good news is that if there's no identifiable convention for this, we get to choose whatever is convenient for how to specify the Sun's location, though!
* I tilt towards `obstime` specifying the sun's location rather than `equinox`. Am I understanding right from @Juanlu001 that this is consistent with the HORIZONS approach?


One other question is how we would want to roll in these changes.  Clearly all the new frames need to go in a new version (3.0), but is it possible/advisable to put *some* of this (most notable I suppose the sun-location-fix) into 2.0.x since it's an actual bug?  I tilt yes but other opinions are useful given how tricky this stuff is...

Juanlu001 commented:
> Suppose I actually care about the actual ecliptic at a specific instant in time because I care about where things are relative to the actual Earth's orbit at that instant. That requires obstime and equinox to be the same. So I believe that means adding @Juanlu001's "Very Optional" case.

Fair enough :)

> I tilt towards obstime specifying the sun's location rather than equinox. Am I understanding right from @Juanlu001 that this is consistent with the HORIZONS approach?

Correct!

>  Clearly all the new frames need to go in a new version (3.0), but is it possible/advisable to put some of this (most notable I suppose the sun-location-fix) into 2.0.x since it's an actual bug?

I leave that decision to the maintainers, but I think it would be very valuable to backport the sun location fix to the current `BarycentricTrueEcliptic` and `HeliocentricTrueEcliptic` regardless of what we actually put in v3.0.0.

Juanlu001 commented:
I intend to work on this by next week, in case you want to make further comments.

Juanlu001 commented:
In fact, I intend to split this in two: one pull request for the non controversial Sun location fix, which can be released as a bug fix, and then another pull request to implement the agreed changes on Ecliptic reference frames.

Juanlu001 commented:
Now I thoroughly went over this again, and realized that actually I didn't add or remove any reference frame: just added an `obstime` attribute to the existing one, which agrees with the latest consensus. So I'm fixing the failing tests here and I think this is ready to merge and backport.

astrofrog commented:
@Juanlu001 - some of the test failures here on Travis are real, but you can ignore the ones that are related to trying to download coordinates/sites.json - these are due to a bug in the pytest remotedata plugin which will be fixed once that plugin has a new version released.

Juanlu001 commented:
Thanks @astrofrog. However, there are some precision issues that seem unrelated to this pull request and that appeared after the merge, for instance:

```
_________________________ test_float_default_precision _________________________
    def test_float_default_precision():
        config = {'pedantic': True}
    
        field = tree.Field(
            None, name='c', datatype='float', arraysize="4",
            config=config)
        c = converters.get_converter(field, config=config)
>       assert (c.output([1, 2, 3, 8.999999], [False, False, False, False]) ==
                '1 2 3 8.9999990000000007')
E       AssertionError: assert '1 2 3 8.999999' == '1 2 3 8.9999990000000007'
E         - 1 2 3 8.999999
E         + 1 2 3 8.9999990000000007
astropy/io/votable/tests/converter_test.py:237: AssertionError
```

Or:

```
                encode('unicode_escape').
>               replace('\\n', '\n'))
E           TypeError: a bytes-like object is required, not 'str'
astropy/io/votable/tests/vo_test.py:166: TypeError
```

I will try to see what happened.

mhvk commented:
The precision issues are for the `numpy-dev` run; those are separate (see #6824; they are allowed to fail, so don't worry about them)

mhvk commented on astropy/coordinates/tests/accuracy/test_ecliptic.py:
```diff
@@ -25,7 +25,7 @@ def test_against_pytpm_doc_example():
     astropy_out = fk5_in.transform_to(pytpm_out)
 
     # we check w/i 1 arcmin because there are some subtle differences in pyTPM's ecl definition
```
Can you update the comment (or maybe just remove it)?

Juanlu001 replied:
Removed!

mhvk commented on astropy/coordinates/tests/test_regression.py:
```diff
@@ -55,7 +55,7 @@ def test_regression_5085():
                            distance=distances, obstime="J2000").cartesian.xyz
     actual_result = coo.transform_to(GCRS(obstime="J2000")).cartesian.xyz
-    assert_quantity_allclose(expected_result, actual_result)
+    assert_quantity_allclose(expected_result, actual_result, rtol=1e-6)
```
Rather than decrease the precision, I think we should just update the expected result.

Juanlu001 replied:
Done.

Juanlu001 commented:
Thanks @mhvk, working on these two bits. Again, the change of coordinate frames will go to another pull request.

Should I add a changelog entry under 2.0.3?

mhvk commented:
I agree this should be 2.0.3 - also, while you're add it, there was a request to add a comment at the erfa function to ensure people do not change it back; see https://github.com/astropy/astropy/pull/6508#issuecomment-326900120

Juanlu001 commented:
> I agree this should be 2.0.3

OK, I added the changelog entry there.

> also, while you're add it, there was a request to add a comment at the erfa function to ensure people do not change it back

Done! :)

mhvk commented:
Great, I'm happy with this, but I'll let @eteq and @StuartLittlefair ping in as well, in particular as a final call for whether we're happy with sticking with the `TrueEcliptic` name.

Juanlu001 commented:
Oops, the Moon computation uses ecliptic coordinates. Fixing it as well.

Juanlu001 commented:
Tests passing :tada: Best use of my Friday evening! :sweat_smile:

StuartLittlefair commented:
This PR looks absolutely AOK to me, apart from the issue about what to do with the ```TrueEcliptic``` coordinate name. 

My personal preference is to leave it as-is. The justification is the obvious health-warning in the documentation, and the coming changes to the ecliptic frame names in v3.0. So it's 👍  from me

astrofrog commented:
@Juanlu001 - can you rebase (no need to squash) on the latest upstream master just to make sure the CircleCI failure goes away? (it should)

Juanlu001 commented:
Done, all lights are green :)

astrofrog commented:
Thanks! I think @eteq or @adrn should merge this.

Juanlu001 commented:
Just as a reminder: I would like to be 100 % sure that this is the way to go before working on the reorganization of ecliptic reference frames while trying to have it ready for inclusion in Astropy 3.0, so if there are still reservations against merging this PR please say so :) Thanks!

eteq commented:
@Juanlu001 - on the Astropy 3.0 changes break out the various frames, I think you should go full steam ahead. I think I also tilt towards `obliquity` as a frame attribute if we can make it work (you might find you need to add a custom frame attribute class if it's supposed to accept either a string *or* a degree value, though...) - it's not the best because I think different obliquity models conceptually map better onto different frames, *but* this is probably one of those "practicality beats purity" cases, given just what you said about how confusing this all is already.

But I still think an actual true (i.e. obstime=equinox) equinox is necessary, because it's the one that maps best *conceptually* to what less-informed people think an equatorial system is.  It may be more a nod to pedagogy than to precision, but that's still useful.

Juanlu001 commented:
Thank you very much @eteq and everybody else that engaged in the discussion! I will be glad to start working on this, hopefully before the end of the year.

Juanlu001 commented:
A year has passed but I did not forget about this :) I hope I can put something together before Astropy 3.2 feature freeze.

pllim commented:
Thank you, @Juanlu001 ! Looking forward to it. Happy New Year!

Juanlu001 commented:
For the interested readers, here's my followup https://github.com/astropy/astropy/pull/8394
